salt/modules/transactional_update.py

```diff
@@ -985,13 +985,40 @@ def call(function, *args, **kwargs):
             else:
                 return local
         except ValueError:
-            return {"result": False, "retcode": 1, "comment": ret_stdout}
+            local = {"result": False, "retcode": 1, "comment": ret_stdout}
+            return local
     finally:
         # Check if reboot is needed
-        if activate_transaction and pending_transaction():
+        if (activate_transaction and pending_transaction()) or (
+            not kwargs.get("test", False) and _user_specified_reboot(local, function)
+        ):
             reboot()
 
 
+def _user_specified_reboot(local, function):
+    if function != "state.highstate" and function != "state.sls":
+        return False
+
+    if not isinstance(local, dict):
+        return False
+
+    explicit_reboot_cmds = set(["reboot", "system.reboot"])
+    explicit_reboot_modules = ["cmd_", "module_"]
```

Comment:
Maybe it's a bit wrong to specify here `_` as a part of the module name, actually `_|-` is used as a separator in the state apply results like you use in the tests:
```
            "cmd_|-reboot_test_|-reboot_|-run": {
                "name": "reboot",
                "changes": {},
                "result": True,
            }
```
I don't know why but actually, `cmd_|-reboot_test_|-reboot_|-run` is a `module name`, `id`, `name`, `fun` separated with `_|-`, so maybe it's better to parse it using `_|-` instead of relying on `|`, but there is a mess that actually either the `name` or `id` can also contain `_|-` and it could add the mess, but even existing code in salt is producing wrong output in such cases.

Reply:
Makes sense,thank you for the suggestion! Implemented.